stellargraph/datasets/dataset_loader.py

```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019-2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Enable easy loading of sample datasets for demonstrations
+"""
+
+import os
+from shutil import unpack_archive
+from urllib.request import urlretrieve
+from typing import List, Optional
+
+
+class DatasetLoader(object):
+    """ A class to download sample datasets.
+    the default path of ~/data can be changed by setting the STELLARGRAPH_DATASETS_PATH environment variable."""
+
+    def __init__(
+        self,
+        name: str,
+        directory_name: str,
+        url: str,
+        url_archive_format: Optional[str],
+        expected_files: List[str],
+        description: str,
+        source: str,
+        data_subdirectory_name: Optional[str] = None,
+    ) -> None:
+        self.name = name
+        self.directory_name = directory_name
+        self.url = url
+        self.url_archive_format = url_archive_format
+        self.expected_files = expected_files
+        self.description = description
+        self.source = source
+        self.data_subdirectory_name = data_subdirectory_name
+
+    @property
+    def base_directory(self) -> str:
+        """Return the path of the data directory for this dataset"""
+        return os.path.join(self._all_datasets_directory(), self.directory_name)
+
+    @property
+    def data_directory(self) -> str:
+        """Return the directory containing the data content files"""
+        if self.data_subdirectory_name is None:
+            return self.base_directory
+        else:
+            return os.path.join(self.base_directory, self.data_subdirectory_name)
+
+    def _create_base_directory(self) -> None:
+        data_dir = self.base_directory
+        if not os.path.exists(data_dir):
+            os.makedirs(data_dir)
+
+    @staticmethod
+    def _all_datasets_directory() -> str:
+        """Return the path of the base directory which contains subdirectories for each dataset."""
+        return os.path.expanduser(
```

Comment:
I think we should strongly consider using something other than `~/data` by default, because that's a directory that a user may well have ideas about the structure of, and already be using a lot. There's somewhat of a standard for a cache directory for various operating systems: e.g. XDG https://wiki.archlinux.org/index.php/XDG_Base_Directory on Linux

https://pypi.org/project/appdirs/ seems to expose this correctly for various platforms, so, if we had it as a dependency, it might be used like 

```python
import appdirs
...

default = appdirs.user_cache_dir("stellargraph", "stellargraph")
return os.getenv("STELLARGRAPH_DATASETS_PATH", default)
```

Reply:
~/data matches old behaviour in StellarGraph but is too generic.  I'm not sure that a standard cache directory is necessarily appropriate here, as it would be difficult for a user to get to.

Reply:
> I'm not sure that a standard cache directory is necessarily appropriate here, as it would be difficult for a user to get to (eg on Mac it might be /Users/x/Library/Application Support/stellargraph and on Linux /home/x/.local/share/stellargraph)

I personally think this is fine because I cannot think of a compelling reason for the user getting to it. Can you give a little more details about why it's important?

> I've updated the path to be ~/stellargraph-datasets (which was already cross-platform due to the use of os.path.expanduser within _all_datasets_directory())
 
This is better since it's less likely to conflict, but it's still junking up a user's home directory for the sake of some demo datasets, that are likely to be only of transient interest to the user (as in, many users will have their own data to work with, and these datasets only serve to demonstrate the library, not solve any real task).

I think it's likely that people will notice and be mildly annoyed: we could reduce some of this annoyance by making it a `.` hidden directory, but I think even that is likely to get people frustrated (e.g. https://github.com/rust-lang/cargo/issues/1734 is a complaint about exactly this issue for Rust's cargo package manager, where it was using `~/.cargo`).

> this avoids dependency on another library too.

That's fair, although we might be able to get away with only installing it for demos and testing (although we'll likely have to put the import inside this function rather than at the top of the file, so it's not evaluated eagerly).

Reply:
I feel a . (hidden) directory is even worse - secretly cluttering the home directory!  The library would be a bit fragile if appdirs wasn't a proper dependancy - passable with documentation but still extra friction (possibly better off as a separate package in that case, though I don't think the current value justifies the extra maintenance overhead of that option)

Reply:
> The current behaviour roughly matches https://scikit-learn.org/stable/datasets/ which downloads to ~/scikit_learn_data/.

Ah, ok, that sounds like a fine precedent, then 👍